content/post/postgres/extension-preloading.md

```diff
@@ -0,0 +1,170 @@
+---
+title: Extension Preloading
+slug: extension-preloading
+date: 2024-07-15T21:03:01Z
+lastMod: 2024-07-15T21:03:01Z
+description: When should a Postgres extension be pre-loaded and when should it not?
+tags: [Postgres, Extensions, Preload]
+type: post
+---
+
+Recently I've been trying to figure out when a Postgres extension module
+should be preloaded. By "extension" I mean, more generally, shared libraries
+provided or used by extensions to Postgres, whether a `CREATE EXTENSION`
+extension written in C or [pgrx] or a [`LOAD`]able module. By "preloaded" I
+mean under what conditions should it be added to one of the [Shared Library
+Preloading] variables:
+
+*   `local_preload_libraries`
+*   `session_preload_libraries`
+*   `shared_preload_libraries`
+
+The answer, it turns out, comes very much down to the extension type.
+
+Normal Extensions
+-----------------
+
+If your extension includes no loadable modules, congratulations! You don't
+have to worry about this question at all.
+
+If your extension's shared library provides functionality only via functions
+called from SQL, you also don't need to worry about this preloading. Custom
+types, operators, and functions generally follow this pattern. The DDL that
+creates the SQL object, such as [`CREATE FUNCTION`], uses the `AS 'obj_file',
+'link_symbol'` syntax to tell that PostgreSQL what library to load when it's
+needed.
+
+Hook Extensions
+---------------
+
+If your extension's shared library makes calls to PostgreSQL without
+PostgreSQL first calling it, then the library must be loaded before it's used.
```

Comment:
First sentence sounds odd - maybe "does something to PostgreSQL before the extension is explictly used"?

Reply:
Yeah, part of the challenge here is that shared libraries/extensions are often _implicitly_ used. Will fiddle with the wording.

Reply:
Changed it to this in efda14c:

> If your shared library needs to perform tasks before PostgreSQL would load it --- or if it would never be loaded implicitly by SQL statements --- then it must be explicitly loaded before it's used.